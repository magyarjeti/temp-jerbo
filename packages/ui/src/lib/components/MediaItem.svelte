<script lang="ts">
	import type { MediaItem, MediaItemSize } from '@packages/api/sdk/v1/schema';
	import JetiIcon from './JetiIcon.svelte';
	import Link from './Link.svelte';
	import { getPhotoswipeContext } from './photoswipe/photoswipe';
	import { onMount } from 'svelte';
	import type { SlideData } from 'photoswipe';

	interface Props {
		wrapperClass?: string | null;
		mediaItem?: MediaItem | string | null;
		enableCaption?: boolean;
		alt?: string | null;
		caption?: string | null;
		showAuthor?: boolean | null;
		preferSize?: MediaItemSize | null;
		enableGalleryView?: boolean;
		class?: string | null;
	}

	let {
		wrapperClass = undefined,
		mediaItem = undefined,
		enableCaption = true,
		alt = undefined,
		caption = undefined,
		showAuthor = false,
		preferSize = undefined,
		enableGalleryView = false,
		...props
	}: Props = $props();

	let mediaSrc = $derived(typeof mediaItem === 'string' ? mediaItem : mediaItem?.url);
	const photoswipeContext = getPhotoswipeContext();
	let el: HTMLElement | undefined = $state();
	let index = $state(0);

	onMount(() => {
		if (enableGalleryView && $photoswipeContext) {
			index = $photoswipeContext.gallery.length;
			const slide: SlideData = {
				alt: alt || '',
				element: el,
			};
			if (typeof mediaItem === 'string') {
				slide.url = mediaItem;
				slide.src = mediaItem;
			} else {
				slide.height = mediaItem?.height;
				slide.width = mediaItem?.width;
				slide.url = mediaItem?.url;
				slide.src = mediaItem?.url;
			}
			$photoswipeContext.gallery = [...$photoswipeContext.gallery, slide];
		}
	});
	const handleImageClick = (ev: Event) => {
		if (enableGalleryView && $photoswipeContext?.openPswp) {
			ev.preventDefault();
			$photoswipeContext.openPswp(index);
		}
	};
	const preferredSize = $derived(
		typeof mediaItem === 'string'
			? null
			: mediaItem?.variations.find((val) => {
					return val.size === preferSize;
				}),
	);
</script>

{#snippet picture()}
	{#if mediaItem}
		{#if typeof mediaItem === 'string'}
			<img class={props.class} src={mediaItem} {alt} />
		{:else if preferSize}
			<img class={props.class} src={preferredSize?.url || mediaItem.url} alt={mediaItem.caption} />
		{:else}
			<picture>
				{#if mediaItem.variations}
					{#each Object.values(mediaItem.variations) as variation}
						<source media="(max-width: {variation.width}px)" srcset={variation.url} />
					{/each}
				{/if}
				<img class={props.class} src={mediaItem.url} alt={mediaItem.caption} />
			</picture>
		{/if}
	{/if}
{/snippet}

<figure bind:this={el} class={wrapperClass}>
	{#if enableGalleryView}
		<Link href={mediaSrc} onclick={handleImageClick}>
			{@render picture()}
		</Link>
	{:else}
		{@render picture()}
	{/if}

	{#if typeof mediaItem !== 'string' && ((mediaItem?.caption && enableCaption) || (mediaItem?.author && showAuthor))}
		<figcaption class="mt-1 flex gap-2">
			<JetiIcon --icon-size={16} class="relative top-[4px] align-top">photo_camera</JetiIcon>
			<div>
				{#if enableCaption}
					<span class="typo-body-small">
						{#if mediaItem?.caption}
							{mediaItem.caption}
						{:else}
							{@html caption}
						{/if}
					</span>
				{/if}
				{#if showAuthor}
					<span class="typo-label-large ml-1">
						<span>Fot√≥: </span>
						{#if mediaItem.sourceLink}
							<Link class="underline hover:no-underline" href={mediaItem.sourceLink}
								>{mediaItem.author}</Link
							>
						{:else}
							<span>{mediaItem.author}</span>
						{/if}
					</span>
				{/if}
			</div>
		</figcaption>
	{/if}
</figure>
