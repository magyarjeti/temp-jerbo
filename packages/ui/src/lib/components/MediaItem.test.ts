import { render } from '@testing-library/svelte';
import { expect, test } from 'vitest';
import type { MediaItem as MediaItemType } from '@packages/api/sdk/v1/schema';

import MediaItem from './MediaItem.svelte';

test('MediaItem with src', () => {
	const { container } = render(MediaItem, {
		props: { src: 'http://example.com' },
	});
	expect(container).toMatchSnapshot();
});

const mediaItem: MediaItemType = {
	id: 'test-id',
	url: 'http://example.com',
	caption: 'test caption',
	author: 'test-author',
	creditType: 'credit type',
	sourceLink: 'http://example.com/source',
	bucket: {
		id: 'test-bucket-id',
		name: 'Test Bucket Name',
		slug: 'test-bucket-slug',
	},
	width: 3456,
	height: 1234,
	variations: [
		{
			url: 'http://example.com/xxs',
			size: 'xxs',
			width: 123,
			height: 45,
		},
		{
			url: 'http://example.com/xs',
			size: 'xs',
			width: 234,
			height: 56,
		},
		{
			url: 'http://example.com/sm',
			size: 'sm',
			width: 456,
			height: 123,
		},
		{
			url: 'http://example.com/md',
			size: 'md',
			width: 567,
			height: 234,
		},
		{
			url: 'http://example.com/lg',
			size: 'lg',
			width: 1234,
			height: 567,
		},
		{
			url: 'http://example.com/xl',
			size: 'xl',
			width: 2345,
			height: 678,
		},
		{
			url: 'http://example.com/xxl',
			size: 'xl',
			width: 4578,
			height: 2345,
		},
	],
};

test('MediaItem with mediaItem', () => {
	const { container } = render(MediaItem, {
		props: { mediaItem },
	});
	expect(container).toMatchSnapshot();
});
