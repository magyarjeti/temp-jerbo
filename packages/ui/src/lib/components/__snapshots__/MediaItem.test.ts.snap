// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MediaItem with mediaItem 1`] = `
<div>
  <figure>
    <picture>
      <source
        media="(max-width: 123px)"
        srcset="http://example.com/xxs"
      />
      <source
        media="(max-width: 234px)"
        srcset="http://example.com/xs"
      />
      <source
        media="(max-width: 456px)"
        srcset="http://example.com/sm"
      />
      <source
        media="(max-width: 567px)"
        srcset="http://example.com/md"
      />
      <source
        media="(max-width: 1234px)"
        srcset="http://example.com/lg"
      />
      <source
        media="(max-width: 2345px)"
        srcset="http://example.com/xl"
      />
      <source
        media="(max-width: 4578px)"
        srcset="http://example.com/xxl"
      />
      
       
      <img
        alt="test caption"
        src="http://example.com"
      />
    </picture>
    
     
    <figcaption
      class="mt-1 inline-flex flex-wrap items-center gap-3"
    >
      <div
        style="display: contents; --icon-size: 16;"
      >
        <span
          class="jeti-icon svelte-1pzkbd4"
        >
          photo_camera
        </span>
      </div>
       
      <span
        class="typo-body-small"
      >
        test caption
      </span>
       
    </figcaption>
  </figure>
</div>
`;

exports[`MediaItem with src 1`] = `
<div>
  <figure>
    <img
      src="http://example.com"
    />
     
  </figure>
</div>
`;
